### PR TITLE
Fix AnomalyDetection Evaluator returning NaN when there are no anomalies

### LINKED ISSUE #2644: AnomalyDetection Evaluator returns NaNs
The `Evaluator` for `AnomalyDetection` returns NaNs for all metrics under some conditions.

The individual metrics shouldn't be returned as a NaN. If something is wrong with the input data, the evaluator should throw an error instead.

Reproduction: Use the MNIST One Class dataset from `tests/data/`, train with default RandomizedPCA parameters, then evaluate the predictions on the training dataset. Observe that the metrics come back as NaN.
```cs
var data = mlContext.Data.ReadFromTextFile<MnistOneClass>(GetDataPath(TestDatasets.mnistOneClass.trainFilename),
    hasHeader: TestDatasets.mnistOneClass.fileHasHeader,
    separatorChar: TestDatasets.mnistOneClass.fileSeparator);

// Create a training pipeline.
var pipeline = mlContext.AnomalyDetection.Trainers.RandomizedPca();

// Train the model.
var model = pipeline.Fit(data);

// Evaulate the model.
var scoredData = model.Transform(data);
var preview = scoredData.Preview();
var metrics = mlContext.AnomalyDetection.Evaluate(scoredData);
```

### PR DESCRIPTION
**Pull request by hichemaichour**
This PR ensures that an exception is thrown when AUC is calculated with a data consisting of only one class.

For anomaly detection, when test data has no anomalies, the returned AUC was NaN because there are no TPs causing the bug mentioned in #2644 

I added Unit Tests to ensure that the exception is thrown

Note: for DrAtK, it looked to me that it is returning NaN by design based on these lines of code:

https://github.com/dotnet/machinelearning/blob/412e1f904c9cd25b80e62bb01390e09360b5b50c/src/Microsoft.ML.Data/Evaluators/AnomalyDetectionEvaluator.cs#L255-L259

Fixes #2644

### CODE REVIEW

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 1]
```diff
@@ -29,33 +22,57 @@ public AnomalyDetectionTests(ITestOutputHelper output) : base(output)
...
             // Evaluate
             var metrics = ML.AnomalyDetection.Evaluate(transformedData, k: 5);
 
-            Assert.Equal(0.98269, metrics.Auc, 5);
+            Assert.Equal(0.98286, metrics.Auc, 5);
             Assert.Equal(0.90000, metrics.DrAtK, 5);
         }
+
+        /// <summary>
+        /// Test anomaly detection when the test data has no anomalies
```
Comment (glebuk) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: >anomalies [](start = 61, length = 9)

nit: comments should end on a period.

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 2]
```diff
@@ -29,33 +22,57 @@ public AnomalyDetectionTests(ITestOutputHelper output) : base(output)
         [Fact]
         public void RandomizedPcaTrainerBaselineTest()
         {
-            var mlContext = new MLContext(seed: 1, conc: 1);
-            string featureColumn = "NumericFeatures";
-
             var reader = new TextLoader(Env, new TextLoader.Options()
```
Comment (glebuk) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: >TextLoader [](start = 29, length = 10)

We should update the tests to use the public API if possible off the MLContext object (see the way ML.AnomalyDetection.Trainers... is used)

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 3]
```diff
@@ -29,33 +22,57 @@ public AnomalyDetectionTests(ITestOutputHelper output) : base(output)
...
                 Separator = "\t",
                 Columns = new[]
                 {
                     new TextLoader.Column("Label", DataKind.R4, 0),
-                    new TextLoader.Column(featureColumn, DataKind.R4, new [] { new TextLoader.Range(1, 784) })
+                    new TextLoader.Column(DefaultColumnNames.Features, DataKind.R4, new [] { new TextLoader.Range(1, 784) })
                 }
             });
 
             var trainData = reader.Read(GetDataPath(TestDatasets.mnistOneClass.trainFilename));
```
Comment (glebuk) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: if you create the reader from the mlcontext you can make it a part of the pipeline.

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 4]
```diff
@@ -29,33 +22,57 @@ public AnomalyDetectionTests(ITestOutputHelper output) : base(output)
...
             Assert.Equal(0.90000, metrics.DrAtK, 5);
         }
+
+        /// <summary>
+        /// Test anomaly detection when the test data has no anomalies
+        /// </summary>
+        [Fact]
+        public void NoAnomalyTest()
+        {
+            var reader = new TextLoader(Env, new TextLoader.Options()
```
Comment (glebuk) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: >TextLoader [](start = 29, length = 10)

duplicate code with method above. Refactor. Perhaps make this a method off the dataset.mnistOneClass?

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs (hunk 1 above)
Comment (hichemaichour) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: Added

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs (hunk 3 above)
Comment (hichemaichour) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: Done

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs (hunk 4 above)
Comment (hichemaichour) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: Moved it into private method

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs (hunk 2 above)
Comment (hichemaichour) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: @glebuk , isn't `ML.AnomalyDetection.Trainers.RandomizedPca()` part of the public API?

I can see that it is from here:

https://github.com/dotnet/machinelearning/blob/4acf5aa97ec4e55b32bf612e551e218b5af2a73a/src/Microsoft.ML.Data/MLContext.cs#L41-L44

Comment (glebuk) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: Yes, the comment was about "new TextLoader(Env,,.." which was not.

---
In reply to: [259996353](https://github.com/dotnet/machinelearning/pull/2692#discussion_r259996353) [](ancestors = 259996353)

Comment (hichemaichour) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: :thumbsup:, reader is derived from mlContext now

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 5]
```diff
@@ -24,39 +18,52 @@ public AnomalyDetectionTests(ITestOutputHelper output) : base(output)
...
 
-            var pipeline = ML.AnomalyDetection.Trainers.RandomizedPca(featureColumn);
+            Assert.Throws<ArgumentOutOfRangeException>(() => ML.AnomalyDetection.Evaluate(transformedData));
+        }
 
-            var transformer = pipeline.Fit(trainData);
-            var transformedData = transformer.Transform(testData);
+        private IDataView DetectAnomalyInMnistOneClass(string trainPath, string testPath)
+        {
+            var reader = ML.Data.CreateTextLoader(new[]
```
Comment (glebuk) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: >new[] [](start = 50, length = 5)

ensure to format the text ctr+k+d

Comment (hichemaichour) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: done, no changes suggested by vs2017 and vscode

File: src/Microsoft.ML.Data/Evaluators/AucAggregator.cs [hunk 6]
```diff
@@ -112,6 +112,8 @@ public override Double ComputeWeightedAuc(out Double unweighted)
                 }
 
                 Contracts.Check(PosSample != null && NegSample != null, "Must call Finish() before computing AUC");
+                Contracts.CheckNonEmpty(PosSample.ToList(), nameof(PosSample), "AUC is not definied when there is no positive class in the data");
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Evaluators/AucAggregator.cs]: Converting this to a full list just to check "not empty" feels too heavy. What about using `.Any()` instead?

Comment (hichemaichour) [reply src/Microsoft.ML.Data/Evaluators/AucAggregator.cs]: True, will change it

Comment (hichemaichour) [reply src/Microsoft.ML.Data/Evaluators/AucAggregator.cs]: done

**Pull request merged**